src/ConstrainedPhotometryModel.cc

```diff
@@ -0,0 +1,213 @@
+}
+
+void ConstrainedPhotometryModel::computeParameterDerivatives(MeasuredStar const &measuredStar,
+                                                             CcdImage const &ccdImage,
+                                                             Eigen::VectorXd &derivatives) const {
+    auto mapping = this->findMapping(ccdImage, "computeParameterDerivatives");
+    mapping->computeParameterDerivatives(measuredStar, measuredStar.getInstFlux(), derivatives);
+}
+
+namespace {
+// Convert photoTransfo's way of storing Chebyshev coefficients into the format wanted by ChebyMap.
+ndarray::Array<double, 2, 2> toChebyMapCoeffs(std::shared_ptr<PhotometryTransfoChebyshev> transfo) {
+    auto coeffs = transfo->getCoefficients();
+    // 4 x nPar: ChebyMap wants rows that look like (a_ij, 1, i, j) for out += a_ij*T_i(x)*T_j(y)
+    ndarray::Array<double, 2, 2> chebyCoeffs = allocate(ndarray::makeVector(transfo->getNpar(), 4));
+    Eigen::VectorXd::Index k = 0;
+    auto degree = transfo->getDegree();
+    for (ndarray::Size j = 0; j <= degree; ++j) {
+        auto const iMax = degree - j;  // to save re-computing `i+j <= degree` every inner step.
+        for (ndarray::Size i = 0; i <= iMax; ++i, ++k) {
+            chebyCoeffs[k][0] = coeffs[j][i];
+            chebyCoeffs[k][1] = 1;
+            chebyCoeffs[k][2] = i;
+            chebyCoeffs[k][3] = j;
+            // std::cout << "idx: " << i << " " << j << " " << k << " " << coeffs[j][i] << " " << std::endl;
+            std::cout << chebyCoeffs[k] << std::endl;
+        }
+    }
+    return chebyCoeffs;
+}
+}  // namespace
+
+std::shared_ptr<afw::image::PhotoCalib> ConstrainedPhotometryModel::toPhotoCalib(
+        CcdImage const &ccdImage) const {
+    std::cout << ccdImage.getName() << std::endl;
+    auto oldPhotoCalib = ccdImage.getPhotoCalib();
+    auto detector = ccdImage.getDetector();
+    auto bbox = detector->getBBox();
+    ChipVisitPhotometryMapping *mapping =
+            dynamic_cast<ChipVisitPhotometryMapping *>(this->findMapping(ccdImage, "toPhotoCalib"));
+    // There should be no way in which we can get to this point and not have a ChipVisitMapping,
+    // so blow up if we don't.
+    assert(mapping != nullptr);
+    auto pixToFocal = detector->getTransform(detector->makeCameraSys(afw::cameraGeom::PIXELS),
+                                             afw::cameraGeom::FOCAL_PLANE);
+
+    // Unravel our chebyshev coefficients to build an astshim::ChebyMap.
+    auto coeff_f = toChebyMapCoeffs(
+            std::dynamic_pointer_cast<PhotometryTransfoChebyshev>(mapping->getVisitMapping()->getTransfo()));
+    // Bounds are the bbox
+    std::vector<double> lowerBound = {static_cast<double>(bbox.getMinX()),
+                                      static_cast<double>(bbox.getMinY())};
+    std::vector<double> upperBound = {static_cast<double>(bbox.getMaxX()),
+                                      static_cast<double>(bbox.getMaxY())};
```

Comment:
I am skeptical of this code. If you are trying to compute values based on pixel position then you want `bboxD = afw::geom::Box2D(bbox)` and then take the limits from `bboxD`. For example: if `bbox = Box2I(Point2I(0, 0), Extent2I(6, 6))` then its limits are `(0, 0)` and `(5, 5)` (since the upper right corner is not included for an int bbox). These indices are also the position of the center of lower left and upper right pixels (after casting to float). But if you want positions that cover edge to edge (the lower left corner of the lower left pixel to the upper right corner of the upper right pixel) you need the range given by `bboxd = Box2D(bbox)`: `(-0.5, -0.5)` and `(5.5, 5.5)`.

Reply:
Yes the code is correct for a native Box2D, but then it doesn't need the `static_cast<double>` and removing the cast would eliminate a source of confusion

Reply:
Ah! Very good point. I'll fix that.